modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala

```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
-                              )
+                    consumerGroupId.flatMap { groupId =>
+                      records.flatTraverse { committable =>
+                        context
+                          .blockOn(blocker) {
+                            F.bracketCase(F.delay(producer.beginTransaction())) { _ =>
+                              committable.records
+                                .traverse(
+                                  KafkaProducer
+                                    .produceRecord(keySerializer, valueSerializer, producer)
+                                )
+                                .flatMap(_.sequence)
+                                .flatTap { _ =>
+                                  F.delay {
+                                    producer.sendOffsetsToTransaction(
+                                      batch.offsets.asJava,
+                                      groupId
+                                    )
+                                  }
+                                }
+                            } {
+                              case (_, ExitCase.Completed) =>
+                                F.delay(producer.commitTransaction())
+                              case (_, ExitCase.Canceled | ExitCase.Error(_)) =>
+                                F.delay(producer.abortTransaction())
                             }
                           }
-                      } {
-                        case (_, ExitCase.Completed) =>
-                          F.delay(producer.commitTransaction())
-                        case (_, ExitCase.Canceled | ExitCase.Error(_)) =>
-                          F.delay(producer.abortTransaction())
                       }
-                    }.flatten
+                    }
                   }
-                }
-              }
 
-              override def toString: String =
-                "TransactionalKafkaProducer$" + System.identityHashCode(this)
+                override def toString: String =
+                  "TransactionalKafkaProducer$" + System.identityHashCode(this)
+              }
             }
           }
+        }
+      }
+
+    private def init(blocker: Blocker)(producer: KafkaByteProducer): F[Unit] =
+      context.blockOn(blocker)(F.delay(producer.initTransactions()))
+
+    private def close(blocker: Blocker)(producer: KafkaByteProducer): F[Unit] =
+      context.blockOn(blocker)(F.delay(producer.close(settings.closeTimeout.asJava)))
+
+    /**
+      * Construct a [[TransactionalKafkaProducer]] which uses a constant ID
+      * for all transactions.
+      *
+      * NOTE: This mode only guarantees exactly-once processing if you have a
+      * single instance in the consumer group. True exactly-once processing is
+      * only guaranteed by running separate producers with unique transactional
+      * IDs per topic/partition processed by the group.
+      */
+    def withConstantId: Resource[F, TransactionalKafkaProducer[F, K, V]] =
```

Comment:
@vlovgr I _really_ want to make this take a `String` parameter to use as the constant ID and have this method inject the appropriate config, to entirely dodge the chance of a runtime error from passing in settings without a transactional ID, but then there isn't much point to having `withTransactionalId` on `ProducerSettings`. What do you think?

It makes even more sense in the dynamic case, where we need an arbitrary prefix for the rebalancing transactional IDs. Pulling the "transactional ID" out of settings and using it as the prefix works, but feels like it could be confusing to users. It also forces our code to deal with the error case when there's no ID present in config, vs. just throwing it at Kafka and catching any errors that bubble up.

Reply:
Agree, let's move transaction-specific configuration to these functions, rather than the Settings.